File: 2023/day21/perl/solution.pl
#!/usr/bin/env perl
# Day 21: Step Counter - Garden plot reachability
# Optimized: removed bigint, use array-based queue with index, packed keys

use strict;
use warnings;

# Parse the input grid and find starting position
sub parse_input {
    my ($filename) = @_;
    open my $fh, '<', $filename or die "Cannot open $filename: $!";
    my @grid;
    my ($start_r, $start_c);
    my $row = 0;
    while (my $line = <$fh>) {
        chomp $line;
        push @grid, $line;
        my $col = index($line, 'S');
        if ($col >= 0) {
            $start_r = $row;
            $start_c = $col;
        }
        $row++;
    }
    close $fh;
    return (\@grid, $start_r, $start_c);
}

# Pack coordinates into a single integer key
# Grid is 131x131, so we need to handle coordinates from about -400 to +400
# Use offset of 500 to make all values positive, then pack as (r+500)*1000 + (c+500)
sub pack_key {
    my ($r, $c) = @_;
    return ($r + 500) * 1000 + ($c + 500);
}

# Part 1: Count cells reachable in exactly 'steps' steps (bounded grid)
sub count_reachable {
    my ($grid, $start_r, $start_c, $steps) = @_;
    my $rows = scalar @$grid;
    my $cols = length $grid->[0];

    # BFS to find minimum steps to each cell
    my %visited;
    my @queue;
    push @queue, [$start_r, $start_c, 0];
    my $qi = 0;
    $visited{pack_key($start_r, $start_c)} = 0;

    while ($qi < @queue) {
        my ($r, $c, $dist) = @{$queue[$qi++]};

        next if $dist >= $steps;

        for my $dir ([-1, 0], [1, 0], [0, -1], [0, 1]) {
            my ($dr, $dc) = @$dir;
            my $nr = $r + $dr;
            my $nc = $c + $dc;

            if ($nr >= 0 && $nr < $rows && $nc >= 0 && $nc < $cols) {
                my $key = pack_key($nr, $nc);
                my $ch = substr($grid->[$nr], $nc, 1);
                if ($ch ne '#' && !exists $visited{$key}) {
                    $visited{$key} = $dist + 1;
                    push @queue, [$nr, $nc, $dist + 1];
                }
            }
        }
    }

    # Count cells reachable in exactly 'steps' steps
    # A cell reachable in d steps can be reached in d+2, d+4, ... steps
    my $target_parity = $steps % 2;
    my $count = 0;
    for my $d (values %visited) {
        if ($d <= $steps && $d % 2 == $target_parity) {
            $count++;
        }
    }
    return $count;
}

# BFS on infinite tiled grid for small step counts
sub count_reachable_infinite_bfs {
    my ($grid, $start_r, $start_c, $steps) = @_;
    my $rows = scalar @$grid;
    my $cols = length $grid->[0];

    my %visited;
    my @queue;
    push @queue, [$start_r, $start_c, 0];
    my $qi = 0;
    $visited{pack_key($start_r, $start_c)} = 0;

    while ($qi < @queue) {
        my ($r, $c, $dist) = @{$queue[$qi++]};

        next if $dist >= $steps;

        for my $dir ([-1, 0], [1, 0], [0, -1], [0, 1]) {
            my ($dr, $dc) = @$dir;
            my $nr = $r + $dr;
            my $nc = $c + $dc;

            my $key = pack_key($nr, $nc);
            # Map to grid coordinates (infinite tiling)
            my $gr = $nr % $rows;
            $gr += $rows if $gr < 0;
            my $gc = $nc % $cols;
            $gc += $cols if $gc < 0;

            my $ch = substr($grid->[$gr], $gc, 1);
            if ($ch ne '#' && !exists $visited{$key}) {
                $visited{$key} = $dist + 1;
                push @queue, [$nr, $nc, $dist + 1];
            }
        }
    }

    # Count cells reachable in exactly 'steps' steps
    my $target_parity = $steps % 2;
    my $count = 0;
    for my $d (values %visited) {
        if ($d <= $steps && $d % 2 == $target_parity) {
            $count++;
        }
    }
    return $count;
}

# Part 2: Count cells reachable in 26501365 steps on infinite grid
# Uses quadratic pattern based on grid structure
sub count_reachable_infinite {
    my ($grid, $start_r, $start_c, $steps) = @_;
    my $rows = scalar @$grid;
    my $size = $rows;  # Grid is square
    my $half = int($size / 2);

    # For small step counts, use direct BFS
    if ($steps <= $size * 2) {
        return count_reachable_infinite_bfs($grid, $start_r, $start_c, $steps);
    }

    # The number of full grid widths we travel
    my $n = int(($steps - $half) / $size);

    # Calculate reachable counts for n=0, 1, 2
    my $y0 = count_reachable_infinite_bfs($grid, $start_r, $start_c, $half);
    my $y1 = count_reachable_infinite_bfs($grid, $start_r, $start_c, $half + $size);
    my $y2 = count_reachable_infinite_bfs($grid, $start_r, $start_c, $half + 2 * $size);

    # Solve for a, b, c using Lagrange interpolation
    # f(x) = y0 + (y1-y0)*x + (y2-2*y1+y0)*x*(x-1)/2
    # a = (y2 - 2*y1 + y0) / 2
    # b = y1 - y0 - a
    # c = y0
    my $a = int(($y2 - 2 * $y1 + $y0) / 2);
    my $b = $y1 - $y0 - $a;
    my $c = $y0;

    return $a * $n * $n + $b * $n + $c;
}

sub part1 {
    my ($grid, $start_r, $start_c) = @_;
    return count_reachable($grid, $start_r, $start_c, 64);
}

sub part2 {
    my ($grid, $start_r, $start_c) = @_;
    return count_reachable_infinite($grid, $start_r, $start_c, 26501365);
}

# Main
my $input_file = $ARGV[0] // "../input.txt";
my ($grid, $start_r, $start_c) = parse_input($input_file);

print "Part 1: ", part1($grid, $start_r, $start_c), "\n";
print "Part 2: ", part2($grid, $start_r, $start_c), "\n";

File: 2023/day21/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
)

type Point struct {
	row, col int
}

func parseInput(filename string) ([]string, Point) {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var grid []string
	var start Point

	scanner := bufio.NewScanner(file)
	row := 0
	for scanner.Scan() {
		line := scanner.Text()
		grid = append(grid, line)
		for col, ch := range line {
			if ch == 'S' {
				start = Point{row, col}
			}
		}
		row++
	}

	return grid, start
}

// countReachable counts cells reachable in exactly 'steps' steps (bounded grid)
func countReachable(grid []string, start Point, steps int) int64 {
	rows, cols := len(grid), len(grid[0])

	// BFS to find minimum steps to each cell
	visited := make(map[Point]int)
	queue := []struct {
		p    Point
		dist int
	}{{start, 0}}
	visited[start] = 0

	dirs := []Point{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]

		if curr.dist >= steps {
			continue
		}

		for _, d := range dirs {
			nr, nc := curr.p.row+d.row, curr.p.col+d.col
			if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
				np := Point{nr, nc}
				if grid[nr][nc] != '#' {
					if _, exists := visited[np]; !exists {
						visited[np] = curr.dist + 1
						queue = append(queue, struct {
							p    Point
							dist int
						}{np, curr.dist + 1})
					}
				}
			}
		}
	}

	// Count cells reachable in exactly 'steps' steps
	// A cell reachable in d steps can be reached in d+2, d+4, ... steps
	targetParity := steps % 2
	var count int64 = 0
	for _, d := range visited {
		if d <= steps && d%2 == targetParity {
			count++
		}
	}

	return count
}

// countReachableInfiniteBFS does BFS on infinite tiled grid for small step counts
func countReachableInfiniteBFS(grid []string, start Point, steps int) int64 {
	rows, cols := len(grid), len(grid[0])

	visited := make(map[Point]int)
	queue := []struct {
		p    Point
		dist int
	}{{start, 0}}
	visited[start] = 0

	dirs := []Point{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

	for len(queue) > 0 {
		curr := queue[0]
		queue = queue[1:]

		if curr.dist >= steps {
			continue
		}

		for _, d := range dirs {
			nr, nc := curr.p.row+d.row, curr.p.col+d.col
			np := Point{nr, nc}

			// Map to grid coordinates (infinite tiling)
			gr := ((nr % rows) + rows) % rows
			gc := ((nc % cols) + cols) % cols

			if grid[gr][gc] != '#' {
				if _, exists := visited[np]; !exists {
					visited[np] = curr.dist + 1
					queue = append(queue, struct {
						p    Point
						dist int
					}{np, curr.dist + 1})
				}
			}
		}
	}

	targetParity := steps % 2
	var count int64 = 0
	for _, d := range visited {
		if d <= steps && d%2 == targetParity {
			count++
		}
	}

	return count
}

// countReachableInfinite counts cells reachable on an infinite tiled grid
func countReachableInfinite(grid []string, start Point, steps int) int64 {
	rows := len(grid)
	size := rows
	half := size / 2

	// For small step counts, use direct BFS
	if steps <= size*2 {
		return countReachableInfiniteBFS(grid, start, steps)
	}

	// The number of full grid widths we travel
	n := int64((steps - half) / size)

	// Calculate reachable counts for n=0, 1, 2 to determine the quadratic
	y0 := countReachableInfiniteBFS(grid, start, half)
	y1 := countReachableInfiniteBFS(grid, start, half+size)
	y2 := countReachableInfiniteBFS(grid, start, half+2*size)

	// Solve for a, b, c using Lagrange interpolation
	// f(x) = y0 + (y1-y0)*x + (y2-2*y1+y0)*x*(x-1)/2
	a := (y2 - 2*y1 + y0) / 2
	b := y1 - y0 - a
	c := y0

	return a*n*n + b*n + c
}

func part1(grid []string, start Point) int64 {
	return countReachable(grid, start, 64)
}

func part2(grid []string, start Point) int64 {
	return countReachableInfinite(grid, start, 26501365)
}

func main() {
	// Get the directory of the executable
	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)
	inputPath := filepath.Join(dir, "..", "input.txt")

	grid, start := parseInput(inputPath)

	fmt.Printf("Part 1: %d\n", part1(grid, start))
	fmt.Printf("Part 2: %d\n", part2(grid, start))
}
